### PR TITLE
Use a more generic layout setting - useIRCLayout → layout

### PR DESCRIPTION
**Pull request by SimonBrandner**
This PR converts the `useIRCLayout` setting to a more generic `layout` setting. [This is a part of the preparations for merging chat-bubbles.](https://github.com/vector-im/element-web/issues/4635#issuecomment-763820551)

### CODE REVIEW

**Comment by t3chguy** [comment]
/me didn't notice it was a draft and got ahead of myself :D

File: src/settings/Layout.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+/* TODO: This should be later reworked into something more generic */
```
Comment (jryans) [inline src/settings/Layout.ts:1]: Add the license header.

File: src/settings/Settings.ts [hunk 2]
```diff
@@ -623,10 +624,9 @@ export const SETTINGS: {[setting: string]: ISetting} = {
         displayName: _td("IRC display name width"),
         default: 80,
     },
-    "useIRCLayout": {
-        supportedLevels: LEVELS_ACCOUNT_SETTINGS,
-        displayName: _td("Enable experimental, compact IRC style layout"),
-        default: false,
+    "layout": {
+        supportedLevels: LEVELS_ROOM_SETTINGS_WITH_ROOM,
```
Comment (jryans) [inline src/settings/Settings.ts:628]: Hmm, so it's per-room here...? The chatter in the bubbles room led me to think we were starting with an account level setting, just like it used to be. Maybe I missed something...?

File: src/components/structures/MessagePanel.js [hunk 3]
```diff
@@ -135,14 +136,13 @@ export default class MessagePanel extends React.Component {
         // whether to show reactions for an event
         showReactions: PropTypes.bool,
 
-        // whether to use the irc layout
-        useIRCLayout: PropTypes.bool,
+        // which layout to use
+        layout: Layout,
```
Comment (jryans) [inline src/components/structures/MessagePanel.js:140]: Hmm, does passing a TypeScript type directly like this actually work...? I assume React is expect one of the `PropTypes` here instead... (Really we should convert this file to TypeScript and do away with the prop types since TS types can be used in place of them, but that should be in a separate PR.)

File: src/components/structures/TimelinePanel.js [hunk 4]
```diff
@@ -111,8 +112,8 @@ class TimelinePanel extends React.Component {
         // whether to show reactions for an event
         showReactions: PropTypes.bool,
 
-        // whether to use the irc layout
-        useIRCLayout: PropTypes.bool,
+        // which layout to use
+        layout: Layout,
```
Comment (jryans) [inline src/components/structures/TimelinePanel.js:116]: See above comments about TS type in prop types.

File: src/components/views/elements/ReplyThread.js [hunk 5]
```diff
@@ -42,7 +43,7 @@ export default class ReplyThread extends React.Component {
         onHeightChanged: PropTypes.func.isRequired,
         permalinkCreator: PropTypes.instanceOf(RoomPermalinkCreator).isRequired,
         // Specifies which layout to use.
-        useIRCLayout: PropTypes.bool,
+        layout: Layout,
```
Comment (jryans) [inline src/components/views/elements/ReplyThread.js:46]: See above comments about TS type in prop types.

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx [hunk 6]
```diff
@@ -28,15 +28,16 @@ import { FontWatcher } from "../../../../../settings/watchers/FontWatcher";
 import { RecheckThemePayload } from '../../../../../dispatcher/payloads/RecheckThemePayload';
 import { Action } from '../../../../../dispatcher/actions';
 import { IValidationResult, IFieldState } from '../../../elements/Validation';
-import StyledRadioButton from '../../../elements/StyledRadioButton';
+//import StyledRadioButton from '../../../elements/StyledRadioButton';
```
Comment (jryans) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:31]: Remove the unused things.

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx [hunk 7]
```diff
@@ -213,15 +214,25 @@ export default class AppearanceUserSettingsTab extends React.Component<IProps, I
         this.setState({customThemeUrl: e.target.value});
     };
 
-    private onLayoutChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
+    /*private onLayoutChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
```
Comment (jryans) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:217]: Needs to be removed I guess?

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx [hunk 8]
```diff
@@ -342,7 +353,7 @@ export default class AppearanceUserSettingsTab extends React.Component<IProps, I
         </div>;
     }
 
-    private renderLayoutSection = () => {
+    /*private renderLayoutSection = () => {
```
Comment (jryans) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:356]: Remove?

File: src/settings/handlers/DeviceSettingsHandler.ts [hunk 9]
```diff
@@ -67,6 +68,13 @@ export default class DeviceSettingsHandler extends SettingsHandler {
             return val['value'];
         }
 
+        // Special case for old useIRCLayout setting
+        if (settingName === "layout") {
+            const settings = this.getSettings() || {};
+            if (settings["useIRCLayout"]) return Layout.IRC;
```
Comment (jryans) [inline src/settings/handlers/DeviceSettingsHandler.ts:74]: Hmm, something seems a bit wrong here, since the old setting is an _account_ setting, but here we're reading from _device_ settings, right?

**Review by jryans — CHANGES_REQUESTED** [review]
Thanks, seems like a good start!

File: src/settings/Settings.ts (hunk 2 above)
Comment (su-ex) [reply src/settings/Settings.ts:628]: Does it hurt having the possibility to set it per room (for example via /-command or console)?
If it's planned to add a per room option later it wouldn't be in the way now and doesn't need to be added later.

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx (hunk 7 above)
Comment (su-ex) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:217]: I need this later. Of course it needs to be changed a little bit but maybe leave it for reference until then?

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx (hunk 8 above)
Comment (su-ex) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:356]: See above.

File: src/settings/handlers/DeviceSettingsHandler.ts (hunk 9 above)
Comment (su-ex) [reply src/settings/handlers/DeviceSettingsHandler.ts:74]: The irc checkbox only ever set the value on the device level.

**Comment by SimonBrandner** [comment]
@jryans Thank you for the review! I've fixed some of the things. Please let me know if the rest is fine this way.

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx (hunk 7 above)
Comment (jryans) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:217]: In my view, it's a bad practice to commit commented out chunks like this. We should instead leverage our tools to help us. For this PR, let's remove the unused things. We can easily recover and modify any bits that are needed later via Git / GitHub.

Comment (SimonBrandner) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:217]: Good point, I'll do that

File: src/settings/Settings.ts (hunk 2 above)
Comment (jryans) [reply src/settings/Settings.ts:628]: Hmm, I believe it's better to make smaller, iterative steps. It doesn't seem like we have resolved yet where it should be per-room, so let's leave this at account settings as it was before. If we get to point of agreeing on a per-room, we'll want to change much more than this line anyway.

File: src/settings/handlers/DeviceSettingsHandler.ts [hunk 10]
```diff
@@ -106,6 +114,18 @@ export default class DeviceSettingsHandler extends SettingsHandler {
             return Promise.resolve();
         }
 
+        // Special case for old useIRCLayout setting
```
Comment (jryans) [inline src/settings/handlers/DeviceSettingsHandler.ts:117]: Since this is a device only level, is there any need to _write_ the old value? It seems like it would only be needed if you are going back in time to old versions, which seems unlikely. (If it were account level and synced to other devices, we would need it, but that doesn't seem to be the case here.)

File: src/settings/handlers/DeviceSettingsHandler.ts (hunk 9 above)
Comment (jryans) [reply src/settings/handlers/DeviceSettingsHandler.ts:74]: Ah okay, makes sense, thanks!

**Review by jryans — CHANGES_REQUESTED** [review]
Thanks for working on this! Still a few things I'd like to see cleaned up, such as removing the commented out blocks.

File: src/settings/handlers/DeviceSettingsHandler.ts (hunk 10 above)
Comment (SimonBrandner) [reply src/settings/handlers/DeviceSettingsHandler.ts:117]: This if statement is here to _delete_ `useIRCLayout`.  
```ts
if (settings["useIRCLayout"]) return Layout.IRC;
``` 
Otherwise, this would always be true.

Comment (jryans) [reply src/settings/handlers/DeviceSettingsHandler.ts:117]: Ah okay, fair enough. I suppose alternatively the read path could always check the modern value first and stop there if it exists, but anyway, this seems okay too.

**Review by jryans — APPROVED** [review]
Thanks for working on this! This part seems ready to go. 😄
